### PR TITLE
Update DerivativeMaterialInterface to match getMaterialProperty

### LINKED ISSUE #5264: Update DerivativeMaterialInterface to match the new getMaterialProperty way
I'm planning to add ```getMaterialPropertyDerivativeByName``` which will have the old behavior and change ```getMaterialPropertyDerivative``` over to accept input parameter keys (with fallback to the *ByName access). This should remove the need to get ```MaterialProperyName``` parameters as strings.

### PR DESCRIPTION
**Pull request by dschwen**
This updated the `DerivativeMaterialInterface` and its `gertMaterialPropertyDerivative` methods to match the style of `getMaterialProperty` introduced with #5218.

I'm also removing the long deprecated (and unused) `getMaterialPropertyPointer` method.

Closes #5264

### CODE REVIEW

**Comment by tonkmr** [comment]
I am happy with these changes.

**Comment by jwpeterson** [comment]
Let's wait to merge this until the issue with Grizzly rolls through the system.

**Comment by dschwen** [comment]
Yes, I'll add the spacing. Btw. Grizzly is green again.

File: framework/include/materials/DerivativeMaterialInterface.h [hunk 1]
```diff
@@ -194,62 +171,138 @@ DerivativeMaterialInterface<T>::getDefaultMaterialProperty(const std::string & p
...
 template<typename U>
-MaterialProperty<U> &
-DerivativeMaterialInterface<T>::declarePropertyDerivative(const std::string &base, const std::vector<std::string> &c)
+bool
+DerivativeMaterialInterface<T>::haveMaterialProperty(const std::string & prop_name)
 {
-  return this->template declareProperty<U>(propertyName(base, c));
+  // Call the correct method to test for material property declarations
+  BlockRestrictable * blk = dynamic_cast<BlockRestrictable *>(this);
+  BoundaryRestrictable * bnd = dynamic_cast<BoundaryRestrictable *>(this);
```
Comment (jwpeterson) [inline framework/include/materials/DerivativeMaterialInterface.h]: If you are going to `dynamic_cast` these, you should at least check the result for NULL before using them.  Even better would be to use 
```
libMesh::cast_ptr<BlockRestrictable *> (this);
```
which does that check for you in debug mode while using a `static_cast` in optimized/oprof mode

File: framework/include/materials/DerivativeMaterialInterface.h [hunk 2]
```diff
@@ -194,62 +171,138 @@ DerivativeMaterialInterface<T>::getDefaultMaterialProperty(const std::string & p
...
+  return ((bnd && bnd->boundaryRestricted() && bnd->template hasBoundaryMaterialProperty<U>(prop_name)) ||
+         (blk && blk->template hasBlockMaterialProperty<U>(prop_name)) ||
+         (this->template hasMaterialProperty<U>(prop_name)));
 }
 
+template<class T>
+template<typename U>
```
Comment (jwpeterson) [inline framework/include/materials/DerivativeMaterialInterface.h]: A one-line `template<class T, typename U>` declaration seems more typical for MOOSE, but we don't have a solid coding convention on that.

File: framework/include/materials/DerivativeMaterialInterface.h (hunk 1 above)
Comment (dschwen) [inline framework/include/materials/DerivativeMaterialInterface.h]: This is code that @aeslaughter added and the cast results **are** checked for NULL (```bnd &&``` in the if).

File: framework/include/materials/DerivativeMaterialInterface.h [hunk 3]
```diff
@@ -194,62 +171,138 @@ DerivativeMaterialInterface<T>::getDefaultMaterialProperty(const std::string & p
...
-MaterialProperty<U> &
-DerivativeMaterialInterface<T>::declarePropertyDerivative(const std::string &base, const std::vector<std::string> &c)
+bool
+DerivativeMaterialInterface<T>::haveMaterialProperty(const std::string & prop_name)
 {
-  return this->template declareProperty<U>(propertyName(base, c));
+  // Call the correct method to test for material property declarations
+  BlockRestrictable * blk = dynamic_cast<BlockRestrictable *>(this);
+  BoundaryRestrictable * bnd = dynamic_cast<BoundaryRestrictable *>(this);
+  return ((bnd && bnd->boundaryRestricted() && bnd->template hasBoundaryMaterialProperty<U>(prop_name)) ||
```
Comment (jwpeterson) [inline framework/include/materials/DerivativeMaterialInterface.h]: Nevermind.

File: framework/include/materials/DerivativeMaterialInterface.h (hunk 2 above)
Comment (dschwen) [inline framework/include/materials/DerivativeMaterialInterface.h]: I can change that along with the ```&``` spacing. It is also the libmesh way apparently.

Comment (dschwen) [inline framework/include/materials/DerivativeMaterialInterface.h]: What is the online line syntax this?
```
template<>
template<typename U>
```

Comment (dschwen) [inline framework/include/materials/DerivativeMaterialInterface.h]: Huh?! This does not compile:
```
DerivativeMaterialInterface.h:258:1: error: too many template parameters in template redeclaration
```
Are you sure those syntaxes are equivalent?!

Comment (dschwen) [inline framework/include/materials/DerivativeMaterialInterface.h]: Pretty sure those are **not** the same thing.

Comment (jwpeterson) [inline framework/include/materials/DerivativeMaterialInterface.h]: I guess they are not the same for partial specialization?  But they must be the same in other cases. I'm googling it...

Comment (jwpeterson) [inline framework/include/materials/DerivativeMaterialInterface.h]: This is the syntax for templated member functions of a template class! Nevermind again.

**Pull request merged**